notebooks/working/Simple_Simulation.md

```diff
@@ -0,0 +1,382 @@
+## Now regress utilities on all the covariates except for travel time and cost. 
+## The coefficient on travel distance now is causal. However, this is not a good model of the output. 
+X = [ 'total_travel_distance',
+       'household_size', 'num_cars', 'cross_bay'
+    ]
+
+utilities_regress_causal = fit_regression(X=X, 
+                                      y = ['utilities'], data=simulation_df, plotting = False)
+utilities_regress_causal.summary()
+```
+
+## What if we don't observe travel distance?
+
+
+### Fit PCA
+
+```python
+import numpy as np
+from numpy.random import randn, rand
+from scipy.optimize import minimize
+import matplotlib.pyplot as plt
+# from nnls import nnlsm_blockpivot as nnlstsq
+import itertools
+from scipy.spatial.distance import cdist
+
+def censored_lstsq(A, B, M):
+    """Solves least squares problem with missing data in B
+    Note: uses a broadcasted solve for speed.
+    Args
+    ----
+    A (ndarray) : m x r matrix
+    B (ndarray) : m x n matrix
+    M (ndarray) : m x n binary matrix (zeros indicate missing values)
+    Returns
+    -------
+    X (ndarray) : r x n matrix that minimizes norm(M*(AX - B))
+    """
+
+    if A.ndim == 1:
+        A = A[:,None]
+
+    # else solve via tensor representation
+    rhs = np.dot(A.T, M * B).T[:,:,None] # n x r x 1 tensor
+    T = np.matmul(A.T[None,:,:], M.T[:,:,None] * A[None,:,:]) # n x r x r tensor
+    try:
+        # transpose to get r x n
+        return np.squeeze(np.linalg.solve(T, rhs), axis=-1).T
+    except:
+        r = T.shape[1]
+        T[:,np.arange(r),np.arange(r)] += 1e-6
+        return np.squeeze(np.linalg.solve(T, rhs), axis=-1).T
+
+
+
+def cv_pca(data, rank, M=None, p_holdout=0.3, nonneg=False, iterations = 1000):
+    """Fit PCA while holding out a fraction of the dataset.
+    """
+
+#     # choose solver for alternating minimization
+#     if nonneg:
+#         solver = censored_nnlstsq
+#     else:
+    solver = censored_lstsq
+
+    # create masking matrix
+    if M is None:
+        M = np.random.rand(*data.shape) > p_holdout
+
+    # initialize U randomly
+    if nonneg:
+        U = np.random.rand(data.shape[0], rank)
+    else:
+        U = np.random.randn(data.shape[0], rank)
+
+    # fit pca/nmf
+    for itr in range(iterations):
+        Vt = solver(U, data, M)
+        U = solver(Vt.T, data.T, M.T).T
+
+    # return result and test/train error
+    resid = np.dot(U, Vt) - data
+    train_err = np.mean(resid[M]**2)
+    test_err = np.mean(resid[~M]**2)
+    return U, Vt, train_err, test_err, M, resid
+```
+
+```python
+X_columns = [
+    'total_travel_time',
+       'total_travel_cost', 
+            ]
+
+
+X = np.array((simulation_df[X_columns] - simulation_df[X_columns].mean())/simulation_df[X_columns].std())
+
+
+# X_raw = np.array([s2,s3]).reshape((1000,2))
+# X =( X_raw - X_raw.mean(axis=0) )/X_raw.std(axis=0)
+
+
+U, Vt, train_err, test_err, M, resid = cv_pca(data=X, rank=2)
+train_err, test_err
+```
+
+### Check PCA
```

Comment:
To check one's PCA (and to check one's models in general), one suggestion I have is to try to always include at least one check that is essentially just a plot of the original data vs the simulated data.

So for the PCA, I might (for each variable) do something like plot the simulated vs observed distributions. For e.g. try looking at a kernel density estimate (or empirical, cumulative distribution function) of the PCA-simulated travel costs vs the observed travel costs.